### PR TITLE
Add support for HerdDB database

### PR DESCRIPTION
**Pull request by eolivelli**
Master Issue: #177 

This patch introduces support for HerdDB as storage:
- new "init database" script
- fix references to 'timestamp' column in SQL, use MySQL escape syntax as 'timestamp' is a reserved word in HerdDB (that in turn uses Apache Calcite as SQL parser)
- fix SQL in "delete" statements to make Apache Calcite happy, as it forbids adding/subtracting unknown datatype (the '?' JDBC parameter is of type UNKNOWN for Calcite at SQL parsing time)
- fixed a query with a "GROUP BY" clause that was not grouping all of the expected columns (Apache Calcite treats that fact as an error)
- added a little example to boot HerdDB in memory-only mode (tested manually on a standalone setup)

It is possible to boot the server inside the application for persistent storage or even for replication.
In case of replication it uses ZooKeeper and BookKeeper.

Side note:
- I had to add some "Maven repositories" to make the build pass locally on my machine

In order to test this branch:
1) change the src/main/resources/application.properties file and check for comments about HerdDB
4) run pulsar manager as usual

### CODE REVIEW

**Comment by eolivelli** [comment]
I had to perform some changes to make the build pass on JDK12 (not needed for HerdDB, but for my local env).

Currently I have a problem with column named 'timestamp'

There are changes to gradle configuration, not needed for HerdDB. It is my own playground

**Comment by eolivelli** [comment]
Found a fix for 'timestamp' -> just use MySQL `timestamp`.

Now I got this other problem about Calcite that is not happy to deal with '-' operator and JDBC parameters

```
### The error may involve io.streamnative.pulsar.manager.mapper.ConsumerStatsMapper.delete-Inline
### The error occurred while setting parameters
### SQL: DELETE FROM consumersStats WHERE ? - ? >= `timestamp`
### Cause: java.sql.SQLException: herddb.model.StatementExecutionException: From line 1, column 34 to line 1, column 38: Cannot apply '-' to arguments of type '<UNKNOWN> - <UNKNOWN>'. Supported form(s): '<NUMERIC> - <NUMERIC>'
'<DATETIME_INTERVAL> - <DATETIME_INTERVAL>'
'<DATETIME> - <DATETIME_INTERVAL>'
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:342)

```

**Comment by eolivelli** [comment]
by the way the Pulsar Manager works

**Comment by tuteng** [comment]
@eolivelli Thank you very much. Is it necessary to create a new file herd for herddb-schema.sql? is it fully compatible with MySQL?

But I threw the following exception when running locally. What do I need to do:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':compileClasspath'.
> Could not find org.herddb:herddb-jdbc:0.12.0-SNAPSHOT.
  Searched in the following locations:
    - file:/Users/tuteng/.m2/repository/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/maven-metadata.xml
    - file:/Users/tuteng/.m2/repository/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/herddb-jdbc-0.12.0-SNAPSHOT.pom
    - file:/Users/tuteng/.m2/repository/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/herddb-jdbc-0.12.0-SNAPSHOT.jar
    - https://jcenter.bintray.com/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/maven-metadata.xml
    - https://jcenter.bintray.com/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/herddb-jdbc-0.12.0-SNAPSHOT.pom
    - https://jcenter.bintray.com/org/herddb/herddb-jdbc/0.12.0-SNAPSHOT/herddb-jdbc-0.12.0-SNAPSHOT.jar
  Required by:
      project :
```

**Comment by eolivelli** [comment]
@tuteng thanks for taking a look.
I have reworked the patch and now it seems to work at 100%.

I am not an user (yet) of Pulsar Manager so I can't tell if every of the features are working as expected.
If an expert of Pulsar Manager could take a tour it would be great.


cc @sijie

**Comment by sijie** [comment]
@eolivelli - @tuteng is the expert of Pulsar Manager. He can help review your code change.

**Comment by eolivelli** [comment]
Cool.
I will be happy to provide working examples for production ready setups.

I still don't have it clear how you a deploying this application.
Is all the state of the application on the sql database? Where do you store sessions?
Is there any way to start more then one inatance of the backend? I don't know SpringBoot at all.

If we start the DB inside the same process you can still run many instances of the backend with HerdDB.

Another point: as far as I can see one Pulsar Manager is able to connect to more than one Pulsar cluster, but there is a 'bookie' option, what is that for?

If you are managing a single Pulsar cluster, with a ZK cluster and  BK cluster you can use that one to support the Herd DB instance

The best setup for standalone (single machine) backend I think in HerdDB case is to run the server inside the backend and provide a directory for data, metadata and journal, without Bookkeeper/ZK.

**Comment by sijie** [comment]
> I still don't have it clear how you a deploying this application.

Pulsar manager comprises of frontend and backend. The frontend is a vuejs application rendering the management and monitoring pages. The backend is a springboot application. You can co-run frontend and backend in one machine or one containerr.

> Is all the state of the application on the sql database? 

Currently we store very minimal informations in SQL database, plus the collected metrics. Most of the restful requests are forwarded back to the real Pulsar brokers.

> Is there any way to start more then one inatance of the backend?

Yes. 

> If we start the DB inside the same process you can still run many instances of the backend with HerdDB.

I am not sure how does it. I would image HerdDB is more like a mysql instance. Then many backend instances will connect to the mysql / herdDB?

> as far as I can see one Pulsar Manager is able to connect to more than one Pulsar cluster, but there is a 'bookie' option, what is that for?

a Pulsar cluster is comprised of brokers, bookies, function workers and many other components. The cluster management and monitoring provides the capability to manage and monitor all the  components.

**Comment by eolivelli** [comment]
Thanks for your clarification @sijie.

I will send links to docs for available deployment options for HerdDB

**Comment by eolivelli** [comment]
@tuteng 
we are going to release HerdDB 0.12.0-SNAPSHOT soon.
I will update this patch and it will download HerdDB dependencies from Maven Central.

Do you have time for a review ?

**Comment by tuteng** [comment]
> @tuteng
> we are going to release HerdDB 0.12.0-SNAPSHOT soon.
> I will update this patch and it will download HerdDB dependencies from Maven Central.
> 
> Do you have time for a review ?

Ok, no problem.

**Comment by eolivelli** [comment]
@tuteng I did another pass, trying to click on every page of the web application.
I found a bunch of issues and committed the fixes

File: src/main/java/io/streamnative/pulsar/manager/mapper/TopicsStatsMapper.java [hunk 1]
```diff
@@ -64,10 +64,10 @@
                 + "sum(msgRateOut) as msgRateOut,"
                 + "sum(msgThroughputOut) as msgThroughputOut,"
                 + "avg(averageMsgSize) as averageMsgSize,"
-                + "sum(storageSize) as storageSize, timestamp FROM topicsStats",
-            "WHERE environment=#{environment} and tenant=#{tenant} and namespace=#{namespace} and timestamp=#{timestamp} and " +
+                + "sum(storageSize) as storageSize, `timestamp` FROM topicsStats",
+            "WHERE environment=#{environment} and tenant=#{tenant} and namespace=#{namespace} and `timestamp`=#{timestamp} and " +
                     "topic IN <foreach collection='topicList' item='topic' open='(' separator=',' close=')'> #{topic} </foreach>" +
-            "GROUP BY cluster, persistent, topic" +
+            "GROUP BY environment, cluster, tenant, namespace, persistent, topic, `timestamp` " +
```
Comment (eolivelli) [inline src/main/java/io/streamnative/pulsar/manager/mapper/TopicsStatsMapper.java]: this "group by" clause was incomplete and Apache Calcite wasn't happy

Comment (tuteng) [reply src/main/java/io/streamnative/pulsar/manager/mapper/TopicsStatsMapper.java]: I fixed it here https://github.com/streamnative/pulsar-manager/pulls

**Comment by eolivelli** [comment]
@tuteng  HerdDB standalone, embedded (same process of the backend)  can handle big databases as this is its primary usecase in production in other projects.

You can evaluate to switch to HerdDB for tests and single instance deployments instead of SQLLite.

Supporting PostGre is good as it is widely used by many companies and the license is better than MySQL.

HerdDB can again be a good alternative to have a replicated (high availability) database, it is really easy to setup replication if you already have a ZK cluster and you can even use your Bookies from Pulsar

**Comment by eolivelli** [comment]
@sijie I will rebase this patch.

**Comment by tuteng** [comment]
@eolivelli  I have done some tests locally and all the features are running well. Thank you very much for your contribution.

However, when I started the service, the log encountered the following errors about HerdDB. I don't know if they have any influence. Can you take a look at them for me?

image

I will consider using it as a test database. Before that, I will first learn about the use of this database.

**Comment by eolivelli** [comment]
The 'ERROR' lines are logged as level SEVERE.
This is expected although it is not very nice with your logging configuration.
I will file an issue

**Comment by sijie** [comment]
>  I will consider using it as a test database.

+1 that's a good idea to start using it as a test database. We can promote HerdDB as one of the databases to use in the documentation.

**Comment by sijie** [comment]
@tuteng @eolivelli if this pull request is almost ready, we can probably move forward with it so that we can start using HerdDB in the tests.

**Comment by eolivelli** [comment]
@sijie awesome !

I will rebase today

**Comment by eolivelli** [comment]
Superseded by #194

**Pull request closed without merging**